src/compiler/crystal/semantic/bindings.cr

```diff
@@ -176,7 +176,18 @@ module Crystal
     end
 
     def set_enclosing_call(enclosing_call)
-      raise "BUG: already had enclosing call" if @enclosing_call
+      current_enclosing_call = @enclosing_call
+      if current_enclosing_call
+        if current_enclosing_call.same?(enclosing_call)
+          # This can happen when a block is typed, and meanwhile a new
+          # generic instance type is created that triggers the block to
+          # be typed again, potentially analyzing a call twice.
+          return
+        else
+          raise "BUG: already had a different enclosing call"
+        end
+      end
+
       @enclosing_call = enclosing_call
```

Comment:
Optional suggestion for more concise code:

```suggestion
      current_enclosing_call = @enclosing_call
      if current_enclosing_call
        # This can happen when a block is typed, and meanwhile a new
        # generic instance type is created that triggers the block to
        # be typed again, potentially analyzing a call twice.
        unless current_enclosing_call.same?(enclosing_call)
          raise "BUG: already had a different enclosing call"
        end
      else
        @enclosing_call = enclosing_call
      end
```

Reply:
I think I had it like this before, but this way ends up setting the instance var again, which is useless. In any case there shouldn't be a performance difference, but I prefer my approach.

Reply:
The assignment is in an else branch. The reason I suggested this is that it makes it very clear that the ivar only gets a value if it didn't have one before.

Reply:
Oh, I see. I was looking at the diff from mobile and it was hard to understand.

That said, the comment now is a bit off. The comment only applies if `same?` is true, but now the comment is outside of that...

Reply:
Well I would interpret the comment as belonging to the then branch of `if current_enclosing_call`.